matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/crypto/InboundGroupSessionStore.kt

```diff
@@ -0,0 +1,96 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.matrix.android.sdk.internal.crypto
+
+import android.util.LruCache
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.launch
+import org.matrix.android.sdk.internal.crypto.model.OlmInboundGroupSessionWrapper2
+import org.matrix.android.sdk.internal.crypto.store.IMXCryptoStore
+import org.matrix.android.sdk.internal.util.MatrixCoroutineDispatchers
+import timber.log.Timber
+import java.util.Timer
+import java.util.TimerTask
+import javax.inject.Inject
+
+/**
+ * Allows to cache and batch store operations on inbound group session store.
+ * Because it is used in the decrypt flow, that can be called quite rapidly
+ */
+internal class InboundGroupSessionStore @Inject constructor(
+        private val store: IMXCryptoStore,
+        private val cryptoCoroutineScope: CoroutineScope,
+        private val coroutineDispatchers: MatrixCoroutineDispatchers) {
+
+    private data class CacheKey(
+            val sessionId: String,
+            val senderKey: String
+    )
+
+    private val sessionCache = object : LruCache<CacheKey, OlmInboundGroupSessionWrapper2>(30) {
+        override fun entryRemoved(evicted: Boolean, key: CacheKey?, oldValue: OlmInboundGroupSessionWrapper2?, newValue: OlmInboundGroupSessionWrapper2?) {
+            if (evicted && oldValue != null) {
+                cryptoCoroutineScope.launch(coroutineDispatchers.crypto) {
+                    Timber.v("## Inbound: entryRemoved ${oldValue.roomId}-${oldValue.senderKey}")
+                    store.storeInboundGroupSessions(listOf(oldValue))
+                }
+            }
+        }
+    }
+
+    private val timer = Timer()
+    private var timerTask: TimerTask? = null
+
+    private val dirtySession = mutableListOf<OlmInboundGroupSessionWrapper2>()
+
+    @Synchronized
+    fun getInboundGroupSession(sessionId: String, senderKey: String): OlmInboundGroupSessionWrapper2? {
+        synchronized(sessionCache) {
+            val known = sessionCache[CacheKey(sessionId, senderKey)]
+            Timber.v("## Inbound: getInboundGroupSession in cache ${known != null}")
+            return known ?: store.getInboundGroupSession(sessionId, senderKey)?.also {
+                Timber.v("## Inbound: getInboundGroupSession cache populate ${it.roomId}")
+                sessionCache.put(CacheKey(sessionId, senderKey), it)
+            }
+        }
+    }
+
+    @Synchronized
+    fun storeInBoundGroupSession(wrapper: OlmInboundGroupSessionWrapper2) {
+        Timber.v("## Inbound: getInboundGroupSession mark as dirty ${wrapper.roomId}-${wrapper.senderKey}")
+        // We want to batch this a bit for performances
+        dirtySession.add(wrapper)
+
+        timerTask?.cancel()
+        timerTask = object : TimerTask() {
+            override fun run() {
+                batchSave()
+            }
+        }
+        timer.schedule(timerTask!!, 2_000)
```

Comment:
I'm always afraid of the use case of user signout during this 2s.

Reply:
yeah, i added a tryOrNull to protect.. but this part is defintly not perfect.
Maybe i could later put that as part of the CryptoStore directly? thus i could save on close